service_exporter/service_exporter.go

```diff
@@ -11,98 +12,124 @@ import (
 
 var log = logging.WithFields(logging.Fields{})
 
+// Metric holds usage and limit desc and values
+type Metric struct {
+	usageDesc *prometheus.Desc
+	limitDesc *prometheus.Desc
+	usage     float64
+	limit     float64
+}
+
 // ServiceQuotasExporter AWS service quotas and usage prometheus
 // exporter
 type ServiceQuotasExporter struct {
-	metricsRegion     string
-	quotasClient      servicequotas.QuotasInterface
-	metricsUsedQuota  map[string]*prometheus.Desc
-	metricsQuotaLimit map[string]*prometheus.Desc
+	metricsRegion  string
+	quotasClient   servicequotas.QuotasInterface
+	metrics        map[string]Metric
+	refreshPeriod  int
+	waitForMetrics chan struct{}
 }
 
 // NewServiceQuotasExporter creates a new ServiceQuotasExporter
-func NewServiceQuotasExporter(region, profile string) (*ServiceQuotasExporter, error) {
+func NewServiceQuotasExporter(region, profile string, refreshPeriod int) (*ServiceQuotasExporter, error) {
 	quotasClient, err := servicequotas.NewServiceQuotas(region, profile)
 	if err != nil {
 		return nil, errors.Wrapf(err, "%w")
 	}
 
+	ch := make(chan struct{})
 	exporter := &ServiceQuotasExporter{
-		metricsRegion:     region,
-		quotasClient:      quotasClient,
-		metricsUsedQuota:  map[string]*prometheus.Desc{},
-		metricsQuotaLimit: map[string]*prometheus.Desc{},
+		metricsRegion:  region,
+		quotasClient:   quotasClient,
+		metrics:        map[string]Metric{},
+		refreshPeriod:  refreshPeriod,
+		waitForMetrics: ch,
 	}
+	go exporter.createQuotasAndDescriptions()
+	go exporter.refreshMetrics()
+
 	return exporter, nil
 }
 
-// Describe writes descriptors to the prometheus desc channel
-func (e *ServiceQuotasExporter) Describe(ch chan<- *prometheus.Desc) {
+func (e *ServiceQuotasExporter) refreshMetrics() {
+	<-e.waitForMetrics
+
+	for {
+		time.Sleep(time.Duration(e.refreshPeriod) * time.Second)
+		e.updateMetrics()
+	}
+}
+
+func (e *ServiceQuotasExporter) updateMetrics() {
 	quotas, err := e.quotasClient.QuotasAndUsage()
 	if err != nil {
 		log.Errorf("Can not retrieve quotas and limits: %s", err)
```

Comment:
Do we need to return this error, is it not possible that we could just silently run into errors retrieving quotas and limits but the exporter would continue running?

Reply:
That's a good point, thanks! I ended up using `Fatalf`, because I think it currently makes sense to just restart the exporter in both cases where we're logging an error.